rustls/src/crypto/mod.rs

```diff
@@ -25,69 +25,58 @@ pub use crate::msgs::handshake::KeyExchangeAlgorithm;
     fn fill_random(buf: &mut [u8]) -> Result<(), GetRandomFailed>;
 
     /// Provide a safe set of cipher suites that can be used as the defaults.
     fn default_cipher_suites() -> &'static [suites::SupportedCipherSuite];
+
+    /// Return a safe set of supported key exchange groups to be used as the defaults.
+    fn default_kx_groups() -> &'static [&'static dyn SupportedKxGroup];
 }
 
-/// An in-progress key exchange over a [SupportedGroup].
-pub trait KeyExchange: Sized + Send + Sync + 'static {
-    /// The supported group the key exchange is operating over.
-    type SupportedGroup: SupportedGroup;
+/// A supported key exchange group.
+///
+/// This has a TLS-level name expressed using the [`NamedGroup`] enum, and
+/// a function which produces a [`ActiveKeyExchange`].
+pub trait SupportedKxGroup: Send + Sync + Debug {
+    /// Named group the SupportedKxGroup operates in.
+    fn name(&self) -> NamedGroup;
 
-    /// Start a key exchange using the [NamedGroup] if it is a suitable choice
-    /// based on the groups supported.
+    /// Start a key exchange.
     ///
     /// This will prepare an ephemeral secret key in the supported group, and a corresponding
-    /// public key. The key exchange must be completed by calling [KeyExchange#complete].
-    ///
-    /// `name` gives the name of the chosen key exchange group that should be used.  `supported`
-    /// is the configured collection of supported key exchange groups. Implementation-specific
-    /// data can be looked up in this array (based on `name`) to allow unconfigured algorithms
-    /// to be discarded by the linker.
+    /// public key. The key exchange can be completed by calling [ActiveKeyExchange#complete]
+    /// or discarded.
     ///
     /// # Errors
     ///
-    /// Returns an error if the [NamedGroup] is not supported, or if a key exchange
-    /// can't be started.
-    fn start(
-        name: NamedGroup,
-        supported: &[&'static Self::SupportedGroup],
-    ) -> Result<Self, KeyExchangeError>;
+    /// This can fail if the random source fails during ephemeral key generation.
+    fn start(&self) -> Result<Box<dyn ActiveKeyExchange>, GetRandomFailed>;
+}
 
+/// An in-progress key exchange originating from a `SupportedKxGroup`.
+pub trait ActiveKeyExchange: Send + Sync {
     /// Completes the key exchange, given the peer's public key.
     ///
-    /// The shared secret is passed into the closure passed down in `f`, and the result of calling
-    /// `f` is returned to the caller.
-    fn complete<T>(self, peer: &[u8], f: impl FnOnce(&[u8]) -> Result<T, ()>) -> Result<T, Error>;
-
-    /// Return the group being used.
-    fn group(&self) -> NamedGroup;
+    /// The shared secret is passed into the [`SharedSecretSink`] for further processing.
+    ///
+    /// This consumes and so terminates the [`ActiveKeyExchange`].
+    fn complete(
+        self: Box<Self>,
+        peer_pub_key: &[u8],
+        sink: &mut dyn SharedSecretSink,
```

Comment:
Hmm, this trait interface feels a little overwrought. Maybe this should just be an `FnOnce` instead?

Reply:
have replaced this with `&mut dyn FnMut(&[u8])` and that's a big improvement (couldn't figure out how to use `FnOnce` without an extra box in a trait method)